src/combinatorics/derangements/basic.lean

```diff
@@ -24,105 +25,134 @@ We also define:
-    rintro ⟨f, _⟩,
-    simp [discard_fixed_pt, equiv.set.compl, derangements,
-          only_possible_fixed_point, not_imp_not] }
-end
+/-- Derangements on a subtype are equivalent to permutations on the original type whose set of fixed
+points is the rest. -/
+protected def subtype_equiv  (p : α → Prop) [decidable_pred p] :
+  derangements (subtype p) ≃ {f : perm α // ∀ a, ¬p a ↔ a ∈ fixed_points f} :=
+calc
+  derangements (subtype p)
+      ≃ {f : {f : perm α // ∀ a, ¬p a → a ∈ fixed_points f} // ∀ a, a ∈ fixed_points f → ¬p a}
+      : begin
+        refine (perm.subtype_equiv p).subtype_equiv (λ f, ⟨λ hf a hfa ha, _, _⟩),
+        { refine hf ⟨a, ha⟩ (subtype.ext _),
+          rw [mem_fixed_points, is_fixed_pt, perm.subtype_equiv] at hfa,
+          dsimp at hfa,
+          rwa equiv.perm.of_subtype_apply_of_mem at hfa },
+        rintro hf ⟨a, ha⟩ hfa,
+        refine hf _ _ ha,
+        rw [mem_fixed_points, is_fixed_pt, perm.subtype_equiv],
+        dsimp,
+        rw [equiv.perm.of_subtype_apply_of_mem _, hfa, subtype.coe_mk],
+      end
+  ... ≃ {f : perm α // ∃ (h : ∀ a, ¬p a → a ∈ fixed_points f), ∀ a, a ∈ fixed_points f → ¬p a}
+      : subtype_subtype_equiv_subtype_exists _ _
+  ... ≃ {f : perm α // ∀ a, ¬p a ↔ a ∈ fixed_points f}
+      : subtype_equiv_right (λ f, by simp_rw [exists_prop, ←forall_and_distrib,
+        ←iff_iff_implies_and_implies])
 
--- TODO elaborator bug means i have to explicitly cast to Type
 /-- The set of permutations that fix at most `a` is equivalent to the sum of:
     - derangements on `α`
     - derangements on `α` minus `a`. -/
-def opfp_equiv_sum_derangements (a : α) :
-  {f : perm α // only_possible_fixed_point f a}
-  ≃ (derangements (({a}ᶜ : set α) : Type _) ⊕ derangements α) :=
-begin
-  let fixes_a := λ f : perm α, f a = a,
-  refine (equiv.sum_compl (λ f : subtype _, fixes_a f.val)).symm.trans (sum_congr _ _),
-  { refine (subtype_subtype_equiv_subtype_inter _ fixes_a).trans
-      (equiv.trans (subtype_equiv_right _) (eofp_equiv_derangements_except_for a)),
-    intro f,
-    exact (eofp_iff_opfp_and_eq a f).symm },
-  { refine (subtype_subtype_equiv_subtype_inter _ (not ∘ fixes_a)).trans (subtype_equiv_right _),
-    intro f,
-    rw mem_derangements_iff_opfp_and_ne }
-end
-
-end fixed_points
+def at_most_one_fixed_point_equiv_sum_derangements [decidable_eq α] (a : α) :
+  {f : perm α // fixed_points f ⊆ {a}} ≃ (derangements ({a}ᶜ : set α)) ⊕ derangements α :=
+calc
+  {f : perm α // fixed_points f ⊆ {a}}
+      ≃ {f : {f : perm α // fixed_points f ⊆ {a}} // a ∈ fixed_points f}
+        ⊕ {f : {f : perm α // fixed_points f ⊆ {a}} // a ∉ fixed_points f}
+      : (equiv.sum_compl _).symm
+  ... ≃ {f : perm α // fixed_points f ⊆ {a} ∧ a ∈ fixed_points f}
+        ⊕ {f : perm α // fixed_points f ⊆ {a} ∧ a ∉ fixed_points f}
+      : begin
+        refine equiv.sum_congr _ _;
+        { convert subtype_subtype_equiv_subtype_inter _ _, ext f, refl }
```

Comment:
this particular lemma was a real PITA. glad to know convert deals with it nicely